skoolkit/basic.py

```diff
@@ -194,24 +254,152 @@ def _get_num_str(self, j):
+        line += '"'
+        return i, line
+
+    def _get_num_array_var(self, i):
+        line = '(Number array) '
+        letter = (self.snapshot[i] & 31) + 96
+        line += "{}(".format(self.text.get_chars(letter))
+        data_length = get_word(self.snapshot, i + 1) - 1
+        dimensions = self.snapshot[i + 3]
+        i += 4
+        dimension_lengths = []
+        for x in range(0, dimensions):
+            dimension_length = get_word(self.snapshot, i)
+            i += 2
+            data_length -= 2
+            line += '{}'.format(dimension_length)
+            if x < (dimensions - 1):
+                line += ','
+        line += ')=['
+        while data_length > 0:
+            line += '{}'.format(get_number(self.snapshot, i))
+            i += 5
+            data_length -= 5
+            if data_length > 0:
+                line += ','
+        line += ']'
+        return i,line
+
+    def _get_long_num_var(self, i):
+        line = '(Number) '
+        letter = (self.snapshot[i] & 31) + 96
+        letters = '{}'.format(self.text.get_chars(letter))
+        i += 1
+        letter = self.snapshot[i]
+        while (letter & 128) == 0:
+            letters += '{}'.format(self.text.get_chars(letter))
+            i += 1
+            letter = self.snapshot[i]
+        letters += '{}'.format(self.text.get_chars(letter & 127))
+        i += 1
+        line += letters + "={}".format(get_number(self.snapshot, i))
+        i += 5
+        return i, line
+
+    def _get_char_array_var(self, i):
+        line = '(Character array) '
+        letter = (self.snapshot[i] & 31) + 96
+        line += "{}$(".format(self.text.get_chars(letter))
+        data_length = get_word(self.snapshot, i + 1) - 1
+        dimensions = self.snapshot[i + 3]
+        i += 4
+        dimension_lengths = []
+        for x in range(0, dimensions):
+            dimension_length = get_word(self.snapshot, i)
+            i += 2
+            data_length -= 2
+            line += '{}'.format(dimension_length)
+            if x < (dimensions - 1):
+                line += ','
+        line += ')=['
+        while data_length > 0:
+            line += '"{}"'.format(self.text.get_chars(self.snapshot[i]))
+            i += 1
+            data_length -= 1
+            if data_length > 0:
+                line += ','
+        line += ']'
+        return i,line
+
+    def _get_control_var(self, i):
+        line = '(FOR control variable) '
+        letter = (self.snapshot[i] & 31) + 96
+        line += "{}=".format(self.text.get_chars(letter))
+        i += 1
+        line += "{} (".format(get_number(self.snapshot, i))
+        i += 5
+        line += 'limit={}, '.format(get_number(self.snapshot, i))
+        i += 5
+        line += 'step={}, '.format(get_number(self.snapshot, i))
+        i += 5
+        line += 'line={}, '.format(get_word(self.snapshot, i))
+        i += 2
+        line += 'statement={})'.format(self.snapshot[i])
+        i += 1
+        return i,line
```

Comment:
This is unnecessarily verbose. It could be reduced to something like this:

    ...
    line += "{}=".format(self.text.get_chars(letter))
    line += "{} (".format(get_number(self.snapshot, i + 1))
    line += 'limit={}, '.format(get_number(self.snapshot, i + 6))
    ...
    return i + 19, line

Same goes for _get_short_num_var().